build-and-deploy-docs/create_mkdocs_config.py

```diff
@@ -0,0 +1,342 @@
+            anchor = get_heading_anchor(heading_content)
+            assert anchor not in anchor_pages
+            anchor_pages[anchor] = pages[-1].get_filename()
+
+        pages[-1].tokens.append(token)
+
+    def sanitize_link(url):
+        """
+        Sanitize a link within the README.
+
+        There are five cases to be handled:
+        1. Links outside of the repository - leave unchanged
+        2. File links within the docs/ folder - rewrite
+        3. Images - copy into docs/ and rewrite
+        4. File links within the repository - redirect to the code browser
+        5. Anchor links - rewrite to reference the newly split pages
+        """
+        link = urlparse(url)
+
+        if link.scheme or link.netloc:
+            # This is a "real" link (https://, ftp://, etc.) - don't touch it
+            pass
+
+        elif link.path:
+            # This is a link to a file on disk
+            resolved_path = Path(readme_file.parent, link.path).resolve()
+
+            # Only mess with paths within the repository
+            if resolved_path.is_relative_to(readme_file.parent):
+                # If the path is already under the docs/ directory, correct it
+                # (the linking document will now be under docs/ as well)
+                if resolved_path.is_relative_to(docs_dir):
+                    link = link._replace(
+                        path=str(resolved_path.relative_to(docs_dir))
+                    )
+
+                # If the link is to an image, copy that image to the docs
+                elif magic.from_file(resolved_path, mime=True) \
+                        in VALID_IMAGE_MIME_TYPES:
+                    output_path = Path(img_dir, resolved_path.name)
+                    shutil.copy2(resolved_path, output_path)
+
+                    link = link._replace(
+                        path=str(output_path.relative_to(docs_dir))
+                    )
+
+                else:
+                    # For everything else, link to the file on GitHub
+                    link = link._replace(
+                        scheme="https",
+                        netloc="github.com",
+                        path=str(Path(
+                            pipeline_repo,
+                            "blob",
+                            os.environ.get("GITHUB_SHA", "main"),
+                            resolved_path.relative_to(readme_file.parent)
+                        ))
+                    )
+
+        elif link.fragment:
+            # This is an anchor link. As we've split the monolithic README into
+            # multiple files, we need to prepend those filepaths.
+            link = link._replace(
+                path=anchor_pages[get_heading_anchor(link.fragment)],
+            )
+
+        return urlunparse(link)
+
+    # Iterate through all tokens (including child tokens) to rewrite links and
+    # copy images into the docs/ folder
+    tokens_to_examine = collections.deque(tokens)
```

Comment:
Is `tokens` flat with no need to recurse? Meaning, can there be, e.g., a header with an anchor that shows up as something like `{'header': {'anchor': ...}}`, or will the tokens always be `['header', 'anchor', ...]`?

Reply:
Unfortunately the tokens are nested:

https://github.com/uclahs-cds/tool-Nextflow-action/blob/ff81b91b5d31f15aed0d7864814ccd1d27812aab/build-and-deploy-docs/create_mkdocs_config.py#L152-L155

Arbitrary tokens are buried in the `inline` block - without this recursing I missed a bunch of links. I could get away with using a list, but the breadth-first-ness of the `deque` is nicer to me.

Reply:
Thanks - you clarified over Teams that the algorithm is walking the tree through a queue and adding children nodes to the end of that queue. Doing this, every node will be visited.